File: netDxf/Entities/PolygonMesh.StoredRecords.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace netDxf.Entities
{
    public partial class PolygonMesh
    {
        private PolygonMeshRecord[] storedVertexRecords;
        private IReadOnlyList<PolygonMeshRecord> storedVertexView = Array.Empty<PolygonMeshRecord>();
        private PolygonMeshRecord storedEndSequence;
        private DxfDocument storedRecordDocument;

        /// <summary>Gets the retained VERTEX identities of a loaded ordinary unsmoothed polygon mesh.</summary>
        /// <remarks>Record indices match the Vertexes array, including its U/V indexing. Authored and smoothed meshes have an empty list.</remarks>
        public IReadOnlyList<PolygonMeshRecord> VertexRecords { get { return this.storedVertexView; } }
        /// <summary>Gets the retained SEQEND, or null when no stored ordinary sequence exists.</summary>
        public PolygonMeshRecord EndSequenceRecord { get { return this.storedEndSequence; } }
        internal bool HasStoredRecords { get { return this.storedEndSequence != null; } }
        internal IEnumerable<PolygonMeshRecord> StoredRecords
        { get { return this.HasStoredRecords ? this.storedVertexRecords.Concat(new[] { this.storedEndSequence }) : Enumerable.Empty<PolygonMeshRecord>(); } }

        internal void SetStoredRecords(DxfDocument document, PolygonMeshRecord[] vertices, PolygonMeshRecord end)
        {
            this.storedRecordDocument = document; this.storedVertexRecords = vertices;
            this.storedVertexView = Array.AsReadOnly(vertices); this.storedEndSequence = end;
            foreach (PolygonMeshRecord record in this.StoredRecords) record.Owner = this;
        }

        internal void BindStoredRecordDocument(DxfDocument document) { this.storedRecordDocument = document; }

        internal void ValidateStoredRecords(DxfDocument document, bool registered)
        {
            if (!this.HasStoredRecords) return;
            if (this.storedRecordDocument != null && !ReferenceEquals(this.storedRecordDocument, document))
                throw new NotSupportedException("Retained polygon mesh records cannot be adopted into another document.");
            if (this.SmoothType != PolylineSmoothType.NoSmooth || this.Vertexes.Length != this.storedVertexRecords.Length)
                throw new NotSupportedException("Changing the retained polygon mesh sequence requires complete schema regeneration.");
            this.ValidateSurface();
            foreach (PolygonMeshRecord record in this.StoredRecords) record.Validate(document, this, registered);
        }

        internal void RejectStoredRecordClone()
        {
            if (!this.HasStoredRecords) return;
            if (this.SmoothType != PolylineSmoothType.NoSmooth)
                throw new NotSupportedException("A retained polygon mesh cannot be cloned after an unsupported smoothing change.");
            this.ValidateSurface();
            if (this.ExtensionDictionary != null || this.PersistentReactors.Count != 0
                || this.XData.Values.SelectMany(data => data.XDataRecord).Any(tag => tag.Code == XDataCode.DatabaseHandle
                    && ulong.Parse((string)tag.Value, System.Globalization.NumberStyles.AllowHexSpecifier, System.Globalization.CultureInfo.InvariantCulture) != 0)
                || this.StoredRecords.Any(record => !record.CanClone()))
                throw new NotSupportedException("Cloning retained polygon mesh records with external, owned or private dependencies requires a complete graph mapping.");
        }

        internal void CopyStoredRecordsTo(PolygonMesh clone)
        {
            if (!this.HasStoredRecords) return;
            this.RejectStoredRecordClone();
            clone.SetStoredRecords(null, this.storedVertexRecords.Select(record => record.CopyForClone()).ToArray(), this.storedEndSequence.CopyForClone());
        }
    }
}

File: javascript/netDxf/Entities/PolygonMesh.StoredRecords.js
// Copyright (c) Daniel Carvajal and netDxf contributors. MIT License; see package LICENSE.
import {ReadOnlyArrayView,HasNonzeroXDataReference} from '../../runtime/StoredRecord.js';
import {NotSupportedException} from '../../runtime/Errors.js';
const storage=new WeakMap();
function state(p){if(!storage.has(p))storage.set(p,{vertices:null,view:ReadOnlyArrayView([]),end:null,document:null});return storage.get(p);}
export function InstallPolygonMeshStoredRecords(Type){
 Object.defineProperties(Type.prototype,{
  VertexRecords:{get(){return state(this).view;}},EndSequenceRecord:{get(){return state(this).end;}},
  HasStoredRecords:{get(){return state(this).end!==null;}},StoredRecords:{get(){return this.HasStoredRecords?[...state(this).vertices,state(this).end]:[];}}
 });
 Type.prototype.SetStoredRecords=function(document,vertices,end){const s=state(this);s.document=document;s.vertices=vertices;s.view=ReadOnlyArrayView(vertices);s.end=end;for(const r of this.StoredRecords)r.Owner=this;};
 Type.prototype.BindStoredRecordDocument=function(document){state(this).document=document;};
 Type.prototype.ValidateStoredRecords=function(document,registered){
  if(!this.HasStoredRecords)return;const s=state(this);
  if(s.document!==null&&s.document!==document)throw new NotSupportedException('Retained polygon mesh records cannot change documents.');
  if(this.SmoothType!==0||this.Vertexes.length!==s.vertices.length)throw new NotSupportedException('Retained polygon mesh topology requires regeneration.');
  this.ValidateSurface();for(const r of this.StoredRecords)r.Validate(document,this,registered);
 };
 Type.prototype.RejectStoredRecordClone=function(){
  if(!this.HasStoredRecords)return;
  if(this.SmoothType!==0)throw new NotSupportedException('A retained polygon mesh cannot be cloned after a smoothing change.');
  this.ValidateSurface();
  if(this.ExtensionDictionary!==null||this.PersistentReactors.Count!==0||HasNonzeroXDataReference(this)||this.StoredRecords.some(r=>!r.CanClone()))throw new NotSupportedException('Cloning retained dependencies requires complete graph mapping.');
 };
 Type.prototype.CopyStoredRecordsTo=function(clone){if(!this.HasStoredRecords)return;this.RejectStoredRecordClone();clone.SetStoredRecords(null,Array.from(state(this).vertices,r=>r.CopyForClone()),state(this).end.CopyForClone());};
}
